src/QMCWaveFunctions/Fermion/MultiDiracDeterminant.cpp

```diff
@@ -529,12 +530,284 @@ void MultiDiracDeterminant::mw_accept_rejectMove(const RefVectorWithLeader<Multi
                                                  int iat,
                                                  const std::vector<bool>& isAccepted)
 {
-  // TODO to be expanded to serve offload needs without relying on calling acceptMove and restore
-  for (int iw = 0; iw < wfc_list.size(); iw++)
+  const int nw = wfc_list.size();
+  // separate accepted/rejected walker indices
+  const int n_accepted = std::count(isAccepted.begin(), isAccepted.begin() + nw, true);
+  const int n_rejected = nw - n_accepted;
+
+  //TODO: can put these in some preallocated work space (reserve up to n_walkers)
+  std::vector<int> idx_Accepted(n_accepted);
+  std::vector<int> idx_Rejected(n_rejected);
+
+  for (int iw = 0, iacc = 0, irej = 0; iw < nw; iw++)
     if (isAccepted[iw])
-      wfc_list[iw].acceptMove(p_list[iw], iat, false);
+      idx_Accepted[iacc++] = iw;
     else
-      wfc_list[iw].restore(iat);
+      idx_Rejected[irej++] = iw;
+
+
+  MultiDiracDeterminant& wfc_leader = wfc_list.getLeader();
+  ParticleSet& p_leader             = p_list.getLeader();
+  const int ndet                    = wfc_leader.getNumDets();
+  const int norb                    = wfc_leader.NumOrbitals;
+  const int nel                     = wfc_leader.NumPtcls;
+  auto& mw_res                      = wfc_leader.mw_res_handle_.getResource();
+
+  const int WorkingIndex = iat - wfc_leader.FirstIndex;
+  assert(WorkingIndex >= 0 && WorkingIndex < wfc_leader.LastIndex - wfc_leader.FirstIndex);
+  assert(p_leader.isSpinor() == wfc_leader.is_spinor_);
+  int handle = 0;
+  for (auto& iacc : idx_Accepted)
+  {
+    auto& wfc = wfc_list.getCastedElement<MultiDiracDeterminant>(iacc);
+    if (wfc.curRatio == ValueType(0))
+    {
+      std::ostringstream msg;
+      msg << "MultiDiracDeterminant::acceptMove curRatio is " << wfc.curRatio << " for walker " << iacc
+          << "! Report a bug." << std::endl;
+      throw std::runtime_error(msg.str());
+    }
+  }
+
+  for (auto& iacc : idx_Accepted)
+  {
+    auto& wfc = wfc_list.getCastedElement<MultiDiracDeterminant>(iacc);
+    wfc.log_value_ref_det_ += convertValueToLog(wfc.curRatio);
+    wfc.curRatio = ValueType(1);
+  }
+
+  // pointers to data for only accepted walkers
+  OffloadVector<ValueType*> psiMinv_temp_acc_deviceptr_list;
```

Comment:
Since there aren't retained I'm not sure why the copy and pointer list building cases aren't fused and these are just in those cases.  Put blocks in the cases and these lists will be constructed only if you are actually using them.

Reply:
yeah, this got a bit messy; I got stuck between not wanting to duplicate code but also not wanting to make the fallthrough logic of the switches hard to follow, so I retained the original logic for the actual copying and used different logic for the pointer lists to avoid duplication (and to only resize the ones I'm actually using), and I ended up with something worse than if I'd just committed to one or the other